### PR TITLE
Add AuthorizeRequest and AuthorizeResponse classes.

### PR DESCRIPTION
**Pull request by LiubovDidkivska**
Add classes with fields needed for decision api
request and responce. Add unit tests.

Relates-To: OLPEDGE-1801

Signed-off-by: Liubov Didkivska <ext-liubov.didkivska@here.com>

### CODE REVIEW

File: olp-cpp-sdk-dataservice-read/include/olp/dataservice/read/AuthorizeRequest.h [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+   *
+   * If operator type not set, type 'and' will be used in request
+   * @see `DecisionApiOperatorType` for more information.
+   *
+   * @return The operator type if set, or boost::none.
+   */
+  inline const boost::optional<DecisionApiOperatorType>& GetOperatorType()
+      const {
+    return operator_type_;
+  }
```
Comment (mykhailo-kuchma) [inline olp-cpp-sdk-dataservice-read/include/olp/dataservice/read/AuthorizeRequest.h:174]: We don't need optional, according to documentation it's AND by default.

File: olp-cpp-sdk-dataservice-read/include/olp/dataservice/read/AuthorizeRequest.h [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+      out << ")";
+    }
+    out << "}";
+    if (GetOperatorType()) {
+      out << "^"
+          << (GetOperatorType().get() == DecisionApiOperatorType::AND ? "AND"
+                                                                      : "OR");
+    }
+    return out.str();
+  }
```
Comment (mykhailo-kuchma) [inline olp-cpp-sdk-dataservice-read/include/olp/dataservice/read/AuthorizeRequest.h:237]: Please move it into CPP file anonymous namespace. (And cleanup includes).

File: olp-cpp-sdk-dataservice-read/include/olp/dataservice/read/AuthorizeRequest.h [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+#include <string>
+#include <vector>
+
+#include <boost/optional.hpp>
+
+#include <olp/dataservice/read/DataServiceReadApi.h>
+
+namespace olp {
+namespace dataservice {
+namespace read {
```
[the remark this answers is not in the document]
Comment (LiubovDidkivska) [reply olp-cpp-sdk-dataservice-read/include/olp/dataservice/read/AuthorizeRequest.h]: moved

File: olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+ */
+class AUTHENTICATION_API AuthorizeResult {
+ public:
+  AuthorizeResult() = default;
+  ~AuthorizeResult() = default;
+
+ private:
+  DecisionType policy_decision_{DecisionType::kDeny};
+  std::string client_id_;
+  std::vector<ActionResult> actions_results_;
```
Comment (mykhailo-kuchma) [inline olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h:110]: public methods first

File: olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+/**
+ * @brief Represents each action/resource pair responce with individual policy
+ * decision for that action DENY or ALLOW.
+ *
+ * @note Data present only if diagnostics flag is true for the DecisionRequest.
+ */
+class AUTHENTICATION_API ActionResult {
+ public:
+  ActionResult() = default;
+  ~ActionResult() = default;
```
Comment (mykhailo-kuchma) [inline olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h:44]: No need for those lines

File: olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+ *
+ * Collected permissions associated with the authenticated user or application,
+ * requested serviceId, and requested contractId For each action/resource pair
+ * in the request returned individual policy decision for that action DENY or
+ * ALLOW.
+ */
+class AUTHENTICATION_API AuthorizeResult {
+ public:
+  AuthorizeResult() = default;
+  ~AuthorizeResult() = default;
```
Comment (mykhailo-kuchma) [inline olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h:105]: No need to have it

File: olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+   *
+   * @return The list of action results.
+   */
+  const std::vector<ActionResult>& GetActionResults() const {
+    return actions_results_;
+  }
+  /**
+   * @brief Sets the list results for each action.
+   *
+   * @param actions vector of ActionResult.
```
Comment (HalynaDumych) [inline olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h:142]:
```suggestion
   * @brief Sets the list of results for each action.
   *
   * @param actions The vector of `ActionResult`.
```

File: olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+   */
+  void SetClientId(std::string client_id) { client_id_ = std::move(client_id); }
+  /**
+   * @brief Gets the list results for each action.
+   *
+   * @note Data present only if diagnostics flag is true for the
+   * DecisionRequest.
+   *
+   * @return The list of action results.
+   */
```
Comment (HalynaDumych) [inline olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h:135]:
```suggestion
  /**
   * @brief Gets the list of results for each action.
   *
   * @note The data is present only if the diagnostics flag is true for
   * `DecisionRequest`.
   *
   * @return The list of action results.
   */
```

File: olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+   * @brief Gets the client id.
+   *
+   * @return The client id.
+   */
+  const std::string& GetClientId() const { return client_id_; }
+  /**
+   * @brief Sets the client id.
+   *
+   * @param The client id.
+   */
```
Comment (HalynaDumych) [inline olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h:126]:
```suggestion
  /**
   * @brief Sets the client ID.
   *
   * @param The client ID.
   */
```

File: olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+   * @brief Sets the overall policy decision.
+   *
+   * @param The decision.
+   */
+  void SetDecision(DecisionType decision) { policy_decision_ = decision; }
+  /**
+   * @brief Gets the client id.
+   *
+   * @return The client id.
+   */
```
Comment (HalynaDumych) [inline olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h:120]:
```suggestion
  /**
   * @brief Gets the client ID.
   *
   * @return The client ID.
   */
```

File: olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+   *
+   * @see `DecisionApiOperatorType` for more information.
+   *
+   * @return The decision based on operator type.
+   */
+  const DecisionType& GetDecision() const { return policy_decision_; }
+  /**
+   * @brief Sets the overall policy decision.
+   *
+   * @param The decision.
```
Comment (HalynaDumych) [inline olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h:113]:
```suggestion
   * @param decision The policy decision.
```

File: olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+ * ALLOW.
+ */
+class AUTHENTICATION_API AuthorizeResult {
+ public:
+  /**
+   * @brief Gets the overall policy decision.
+   *
+   * @see `DecisionApiOperatorType` for more information.
+   *
+   * @return The decision based on operator type.
```
Comment (HalynaDumych) [inline olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h:107]:
```suggestion
   * @return The decision based on the operator type.
```

File: olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+
+/**
+ * @brief A model used to represent policy decision for a given request context
+ * against a HERE Service.
+ *
+ * Collected permissions associated with the authenticated user or application,
+ * requested serviceId, and requested contractId For each action/resource pair
+ * in the request returned individual policy decision for that action DENY or
+ * ALLOW.
+ */
```
Comment (HalynaDumych) [inline olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h:99]:
```suggestion
/**
 * @brief Represents the policy decision for a given request context
 * against the HERE Service.
 *
 * Collects all permissions associated with the authenticated user or
 * application, requested service ID, and requested contract ID. For each
 * action-resource pair in the request, this class determines an individual
 * policy decision: DENY or ALLOW.
 */
```

File: olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+   * If ALL considered permissions for the action have an effect of ALLOW, the
+   individual policy decision for the action is ALLOW
+
+   * @return The list permitions.
+   */
+  const std::vector<Permisions>& GetPermitions() const { return permisions_; }
+  /**
+   * @brief Sets the list of permitions.
+   *
+   * @param permitions vector of pair action/decision.
```
Comment (HalynaDumych) [inline olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h:80]:
```suggestion
   * @param permitions The vector of the action-decision pair.
```

File: olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+   and resource.
+   * If the action matches the action in the permission and the resource
+   matches* the resource in the permission, consider the permission
+   * If ANY considered permission for the action has an effect of DENY, the
+   individual policy decision for the action is DENY
+   * If ALL considered permissions for the action have an effect of ALLOW, the
+   individual policy decision for the action is ALLOW
+
+   * @return The list permitions.
+   */
```
Comment (HalynaDumych) [inline olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h:75]:
```suggestion
   * @note Algorithm of evaluating each permission in the set against the action
   and resource.   
   * If the action matches the action in the permission, and the resource
   matches the resource in the permission, consider the permission.
   * If ANY considered permission for the action has results in DENY, the
   individual policy decision for the action is DENY.
   * If ALL considered permissions for the action results in ALLOW, the
   individual policy decision for the action is ALLOW.

   * @return The list of permitions.
   */
```

File: olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h [hunk 16]
```diff
@@ -28,56 +28,60 @@ namespace olp {
...
    */
   void SetDecision(DecisionType decision) { decision_ = decision; }
+
   /**
-   * @brief Gets the list of permitions, which are used to evaluate against the
+   * @brief Gets the list of permitions that are evaluated against the
    action and resource.
    *
    * @note Algorithm of evaluating each permission in the set against the action
    and resource.
```
Comment (HalynaDumych) [inline olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h]:
```suggestion
   * @note The algorithm of evaluating each permission in the set against the action
   and resource:
```

File: olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h [hunk 17]
```diff
@@ -28,56 +28,60 @@ namespace olp {
...
-   individual policy decision for the action is ALLOW
-
-   * @return The list permitions.
+   * If the action matches the action in the permission, and the resource
+   matches the resource in the permission, consider the permission.
+   * If ANY considered permission for the action has results in DENY, the
+   individual policy decision for the action is DENY.
+   * If ALL considered permissions for the action results in ALLOW, the
+   individual policy decision for the action is ALLOW.
+   * @return The list of permitions.
```
Comment (HalynaDumych) [inline olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h:77]:
```suggestion
   * * If the action matches the action in the permission, and the resource
   matches the resource in the permission, consider the permission.
   * * If ANY considered permission for the action results in DENY, the
   individual policy decision for the action is DENY.
   * * If ALL considered permissions for the action result in ALLOW, the
   individual policy decision for the action is ALLOW.
   *
   * @return The list of permitions.   
```

File: olp-cpp-sdk-authentication/src/AuthorizeRequest.cpp [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+ */
+
+#include "olp/authentication/AuthorizeRequest.h"
+
+#include <sstream>
+
+namespace olp {
+namespace authentication {
+
+std::string AuthorizeRequest::CreateKey() const {
```
Comment (r0busta) [inline olp-cpp-sdk-authentication/src/AuthorizeRequest.cpp:27]: The method is non-trivial and it'd be nice to have a few unit tests for it.

File: olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeRequest.h [hunk 19]
```diff
@@ -0,0 +1,233 @@
...
+   * @brief Gets the operator type.
+   *
+   * If the operator type is not set, the 'and' type is used in the request.
+
+   * @see `DecisionApiOperatorType` for more information.
+   *
+   * @return The operator type.
+   */
+  inline DecisionApiOperatorType GetOperatorType()
+      const {
```
Comment (andescu) [inline olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeRequest.h:178]: This doesn't seem correctly formatted.

File: olp-cpp-sdk-authentication/src/AuthorizeRequest.cpp [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+    if (actions.second) {
+      out << ", " << actions.second.get();
+    }
+    out << ")";
+  }
+  out << "}";
+  out << "^"
+        << (GetOperatorType() == DecisionApiOperatorType::kAnd ? "AND"
+                                                              : "OR");
+  return out.str();
```
Comment (andescu) [inline olp-cpp-sdk-authentication/src/AuthorizeRequest.cpp:45]: This is too much. The sole purpose of the key is to have nice print log. I would say service_id and contract_id are totally sufficient.

File: olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeRequest.h [hunk 21]
```diff
@@ -0,0 +1,233 @@
...
+   *
+   * @param[in] action The action that is used to determine an individual policy
+   * decision.
+   * @param[in] resource The resource that is used to request the decision for
+   * the action.
+   *
+   * @return A reference to the updated `DecisionRequest` instance.
+   */
+  inline AuthorizeRequest& WithAction(
+      std::string action, boost::optional<std::string> resource = boost::none) {
```
Comment (andescu) [inline olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeRequest.h:163]: Are you sure that the resource should be optional? Can it not be just a nomal string with users responsability to leave empty or fill with value.

File: olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeRequest.h [hunk 22]
```diff
@@ -0,0 +1,233 @@
...
+   * decision of ALLOW, the overall policy decision returned
+   * in the response is ALLOW.
+   * If the operator is 'and' or missing (defaults to 'and'),
+   * one of the following algorithms is applied:
+   * * If ANY actions have an individual policy decision of DENY,
+   *  the overall policy decision returned in the response is DENY.
+   * * If ALL actions have an individual policy decision of ALLOW,
+   * the overall policy decision returned in the response is ALLOW.
+   */
+  enum class DecisionApiOperatorType { kAnd, kOr };
```
Comment (andescu) [inline olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeRequest.h:79]:
```suggestion
  enum class DecisionOperatorType { kAnd, kOr };
```

File: olp-cpp-sdk-authentication/include/olp/authentication/Types.h [hunk 23]
```diff
@@ -43,5 +44,7 @@ using IntrospectAppResponse = Response<IntrospectAppResult>;
 /// Called when the user introspect app request is completed.
 using IntrospectAppCallback = Callback<IntrospectAppResult>;
 
+/// The decision API response alias.
+using AuthorizeResponse = Response<AuthorizeResult>;
```
Comment (andescu) [inline olp-cpp-sdk-authentication/include/olp/authentication/Types.h:48]: You are missing the Callback<AuthorizeResult>

File: olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeRequest.h (hunk 21 above)
Comment (LiubovDidkivska) [reply olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeRequest.h]: changed to empty string

File: olp-cpp-sdk-authentication/src/AuthorizeRequest.cpp (hunk 18 above)
Comment (LiubovDidkivska) [reply olp-cpp-sdk-authentication/src/AuthorizeRequest.cpp:27]: Please look https://github.com/heremaps/here-olp-sdk-cpp/pull/792/files/b04c47be5011cbe66ba5fb8379e729b4f4191bf8#diff-d1d1293083a502f525306c203f4bc75f, there is some checks for CreateKey

Comment (r0busta) [reply olp-cpp-sdk-authentication/src/AuthorizeRequest.cpp:27]: There's just one check. It doesn't check the `if (GetContractId())` branch or `AND` case.

Comment (LiubovDidkivska) [reply olp-cpp-sdk-authentication/src/AuthorizeRequest.cpp:27]: File was updated https://github.com/heremaps/here-olp-sdk-cpp/pull/792/commits/b78a32acf765d415c043697596f75e533eeda2bd.

File: olp-cpp-sdk-authentication/src/AuthorizeRequest.cpp [hunk 24]
```diff
@@ -0,0 +1,37 @@
...
+ */
+
+#include "olp/authentication/AuthorizeRequest.h"
+
+#include <sstream>
+
+namespace olp {
+namespace authentication {
+
+std::string AuthorizeRequest::CreateKey() const {
```
Comment (andescu) [inline olp-cpp-sdk-authentication/src/AuthorizeRequest.cpp:27]: You should be able to inline this inside the header and save the cpp file.

Comment (LiubovDidkivska) [reply olp-cpp-sdk-authentication/src/AuthorizeRequest.cpp:27]: @mykhailo-kuchma doesn't like additional includes as sstream in header file(https://github.com/heremaps/here-olp-sdk-cpp/pull/792#discussion_r411947706). And suggest to remove this function at all. What do you think, should we remove it?

Comment (andescu) [reply olp-cpp-sdk-authentication/src/AuthorizeRequest.cpp:27]: We used the inline approach in dataservice-read.
But it doesn't matter very much. I guess it is fine like this also.

**Pull request merged**